### ISSUE TITLE
Questions about normal settings

### ISSUE THREAD

**Issue by yinyunie** [maturk/dn-splatter#9]
Hi,

Thanks a lot for your great work!

I have some questions about the normals' coordinate system:

1. I noticed that the output normals files under the `normals_from_pretrain` folder generated from `normals_from_pretrain.py` (using [omnimodel](https://github.com/EPFL-VILAB/omnidata)) are under the `opencv` system, but the default `normal_format` in `*_dataparser.py` is `opengl`. Should I append the config with `... "--normal_format", "opencv"` or keep the default `opengl`?

2. In `dn_dataset.py` [[L207](https://github.com/maturk/dn-splatter/blob/925d985ecb64c6544ecde936de5e8a0dd06e5399/dn_splatter/data/dn_dataset.py#L207)], I found if `normal_format == "opengl"`, the normals under `opengl` will be converted to `opencv` for supervision, then I assume the predicted normals should also be under "opencv" system for loss calculation.

3. However, in `get_loss_dict(*)` in creating GT normals from depth [Link](https://github.com/maturk/dn-splatter/blob/925d985ecb64c6544ecde936de5e8a0dd06e5399/dn_splatter/dn_model.py#L740C1-L756C48), I found the `gt_normal` is created under "opengl" system. Which is conflicting with 2.
```
                gt_normal = normal_from_depth_image(
                    depths=depth_out.detach(),
                    fx=self.camera.fx.item(),
                    fy=self.camera.fy.item(),
                    cx=self.camera.cx.item(),
                    cy=self.camera.cy.item(),
                    img_size=(self.camera.width.item(), self.camera.height.item()),
                    c2w=torch.eye(4, dtype=torch.float, device=depth_out.device),
                    device=self.device,
                    smooth=False,
                )
                gt_normal = gt_normal @ torch.diag(
                    torch.tensor(
                        [1, -1, -1], device=depth_out.device, dtype=depth_out.dtype
                    )
                )
                gt_normal = (1 + gt_normal) / 2
```

Looking forward to your reply.

Best,
Yinyu

**Comment by maturk** [comment]
Hi @yinyunie , thank you for the question. It is an important one and something that I struggled myself with for quite some time. I hope that I have not actually made an error where I have mistakenly named OpenGL as OpenCV coordinate system. This is my current understanding of the coordinate frames regarding normals as well as how I render/convert normals in this project:

1) From the best of my understanding, the output of Omnidata model is in an OpenGL coordinate system. Since I have only tested one monocular-normal network (Omnidata) I have made all dataparsers have the "--normal-format opengl" set as default.
2) Most important thing in my codebase is that I render and generate all normals and their losses in the camera frame in the OpenCV convention. This is because I am most familiar with the Open3D visualization of normals (they use OpenCV camera coordinates), so this consistency in visualization was vital for me in the beginning for debugging resons. Therefore, the normals processed by `dn_dataset.py` [here](https://github.com/maturk/dn-splatter/blob/925d985ecb64c6544ecde936de5e8a0dd06e5399/dn_splatter/data/dn_dataset.py#L207) undergo a coordinate transformation from OpenGL to OpenCV. These are then loaded by the model, and all losses and also visualizations in the viser client are done in this OpenCV coordinate system. 
3) Here is a small visualization that has helped me distinguish the differences. From the best of my understanding the left image normals are in OpenGL and the right are in OpenCV. 

![image (8)](https://github.com/maturk/dn-splatter/assets/30566358/4d903956-34a0-4138-8415-721b6c9cad96)

4) Now regarding the case where you choose to use a different monocular normal network, e.g. Dsine, which already outputs normals in the OpenCV coordinate system, then it is vital to set the "--normal-format opencv" flag to bypass any coordinate transformations done in `dn_dataset.py. ` I know the user @pablovela5620 has been testing out Dsine recently, and I hope to also add his work into this project and also to update the readme to make things more clear. 

5) Another important convention is that currently I load all normals as `.png` images in the range [0,1]. I also compute all losses in this range as well. I effectively treat normals just as images so you will see some conversions from the usual [-1,1] range to [0,1] in the code. 

Please let me know if you have further questions and doubts. This is someting that I spent quite some time debugging and this is my current best understanding. Thanks!

**Comment by yinyunie** [comment]
Thanks a lot! It helps a lot.

Here is my experiment to demonstrate why I think the normals from `normals_from_pretrain.py` is under `opencv` system.

![Screenshot from 2024-04-12 13-48-41](https://github.com/maturk/dn-splatter/assets/25686434/2a63bb4e-97e3-4487-b7b3-5747e4ea73bf)

Here is the input image and the normal map generated from `normals_from_pretrain.py` by 
`python dn_splatter/scripts/normals_from_pretrain.py --data-dir data --resolution low --img-dir-name test_imgs`.
It saved a `.npy` file and a `.png` file. Multiply the matrix in `.npy` by 255 will reproduce the `.png` file.

Convert the normal matrix into camera system:
```
>>> normal = np.load("data/normals_from_pretrain/test_img.npy")
>>> normal = normal * 2 - 1
>>> print(normal.min())
-1.0
>>> print(normal.max())
1.0
```
Check the normal vector of the floor:
```
>>> print(normal.shape)
(3, 958, 1588)
>>> print(normal[:, -1, 0])  # [-1, 0] is the index of bottom left corner on the image.
[ 0.00621688 -0.9692557  -0.11855131]
```
Check the normal vector of the ceiling:
```
>>> print(normal[:, 0, 1588//2]) # [0, 1588//2] is the index of top center of the image.
[0.01825583 0.98115635 0.0951637 ]
```
Check the normal vector of the front wall:
```
>>> print(normal[:, 958//2, -1]) # [958//2, -1] is the middle of right side of the image.
[-0.51757413  0.07216501 -0.8123885 ]
```

It explains well the "opencv" system like below:
![Screenshot from 2024-04-12 14-13-05](https://github.com/maturk/dn-splatter/assets/25686434/18404c35-bede-4680-a54b-6ce51c0af306)

I am not sure if I missing something. 

Thanks a lot.

**Comment by pablovela5620** [comment]
I think you're right that the OmniData output is encoded in opencv format (Right Down Forward | X Y Z) based both on your experiments and this comment here from the original omnidata authors
https://github.com/EPFL-VILAB/omnidata/issues/13#issuecomment-1209543393
it may just be a notation error in the dnsplatter repo

**Comment by altaykacan** [comment]
@maturk thank you for your great work and your engagement with helping out with issues! I am working on a similar project so your work is super helpful!

I just wanted to chime in and hopefully help out with the discussion. I think the notations you provided in your examples are flipped. If I understand correctly you mean the left image is OpenGL and the right image is in OpenCV coordinates.

From what I found online, the coordinate systems of the respective cameras look like this for both frameworks ([source](https://stackoverflow.com/questions/44375149/opencv-to-opengl-coordinate-system-transform)):

![image](https://github.com/maturk/dn-splatter/assets/33330747/22a0bcb1-3c70-4013-be14-8bf15b59bfbd)

I made this quick and dirty drawing to clarify how both systems would look like in the image you've provided. I drew the OpenCV system in red (y pointing down) and the OpenGL system in yellow (y pointing up):

![Screenshot 2024-04-13 at 16 40 00](https://github.com/maturk/dn-splatter/assets/33330747/6a4ceb19-be67-44b8-8d2c-914963cdf37c)

Looking at the ground, the normal vector of the surface should be pointing up. This corresponds to negative y coordinates in the OpenCV system and positive y ones in the OpenGL system. The x and z coordinates of this upwards pointing vector should be zero for both systems. Assuming the normals are represented with float values between [-1,1] in the camera coordinate frame and the mapping is `(x,y,z) --> (r,g,b)`  and `[-1, 1] --> [0, 255]`. I used a basic RGB slider to see which colors we'd get after mapping the normal values.

A positive y value would correspond to this color (red and blue values should be roughly in the middle since the x and z values should be zero) :

![image](https://github.com/maturk/dn-splatter/assets/33330747/fb214b37-b6b1-4cfb-949c-b22a1da3e77d)

and a negative y value would be this color:

![image](https://github.com/maturk/dn-splatter/assets/33330747/e681463c-c2f4-412f-9591-22b4bf1d4bdd)

Since the color of the floor in the left image matches this dark purple-color and that corresponds to a negative y value as would be represented in the OpenCV system. So the left image should not be in the OpenGL system like you mentioned. I am not sure whether this is a typo or a misunderstanding from my side but I just wanted to share my thought process.

Hopefully this helps someone and we can clarify the issue!

**Comment by altaykacan** [comment]
> I think you're right that the OmniData output is encoded in opencv format (Right Down Forward | X Y Z) based both on your experiments and this comment here from the original omnidata authors [EPFL-VILAB/omnidata#13 (comment)](https://github.com/EPFL-VILAB/omnidata/issues/13#issuecomment-1209543393) it may just be a notation error in the dnsplatter repo

@pablovela5620 thank you for linking the comment I think it's pretty helpful! From what the author says there _"[...] a vector of (0,0,1) faces towards the camera."_ what I understand is that omnidata is actually using the OpenGL system as @maturk mentions but as I explained above I think images he gave as examples are flipped. 

A vector of (0,0,1) would be this color:

![image](https://github.com/maturk/dn-splatter/assets/33330747/ece18317-5978-4b04-b472-fc52d5d8762a)

and looking at the right image from @maturk 's answer I'd say the wall facing the camera has this color:

![image](https://github.com/maturk/dn-splatter/assets/33330747/0b85c514-935c-4dc0-bbdc-e440054e3a1a)

Since the normal of that wall is roughly facing the camera and is associated with a positive z-value, the normals would actually be in the OpenGL system.

This is my understanding of the issue and I hope this helps out! I'd appreciate if you notice anything wrong in my reasoning and let me know!

**Comment by pablovela5620** [comment]
This is great and has helped me clear up a lot! Similarly to @altaykacan and @yinyunie, I did some experiments to verify the data range + camera coordinate systems of the surface normals. 

To start, I logged the output from the omnidata surface normal on some data and got the following. The only post processing I performed was converting from a tensor to numpy and reshaping back to the original size (since omni data only accepts 384x384 images)
![omni_datainfo](https://github.com/maturk/dn-splatter/assets/25287427/3b8188ac-d9e2-4542-964d-36a9235acae8)

The output has a range between 0-1, and as @maturk mentioned is traditionally in the range of -1 to 1 to have the appropriate values (though for the actual network training, it sounds like the values need to be between 0-1 which is the mistake I was making on my side when implementing dsine!)

For the dsine network, the output appears to be between -1 and 1 (as compared to omnidata). Since It looks like omnidata is actually in OpenCV coordinates, I wanted to verify that the Dsine network is in OpenGL coordinates. I took some images of logged normals and verified as much
# Here is a surface that should be in the positive x direction (Red channel in this case)
![dsine_x](https://github.com/maturk/dn-splatter/assets/25287427/1a3c4d4c-89dc-4238-8926-565570b4f024)

# Here is a surface that should be in the positive y direction (Green channel in this case)
![dsine_y](https://github.com/maturk/dn-splatter/assets/25287427/0567ee03-d707-4157-85ac-2eae882aaef1)

# Here is a surface that should be in the positive z direction (Blue channel in this case)
![dsine_z](https://github.com/maturk/dn-splatter/assets/25287427/9c2dbaf8-3650-4242-bfa2-1a3c2ade4327)



# TLDR I think that as mentioned the convention naming is actually incorrect and the output of omnidata is the OpenCV convention (X ->Right | Y -> Down | Z -> Forward) (RDF)while the convention that is shown in the dn-splatter viewer + dsine network is in OpenGL (X -> Right | Y-> Up | Z -> Back) (RUB)
